### PR TITLE
create upload action

### PR DESCRIPTION
**Pull request by MoranLi**
## Description

Create Upload state and action
Create required mutation types and mutations
Create getters

## Related Issue or Trello Card

[Trello card](https://trello.com/c/KoJZRLbt/8-create-service-for-uploading-files-to-firebase-cloud-storage)

## Motivation and Context

To be able to upload an image file to the firebase storage bucket

## How Has This Been Tested?

- [x] Unit test
- [ ] NightWatch test
- [ ] Coverage testing
- [x] Manual testing

## Test Cases and test paths

For automatic unit testing, testing file is in the unit/specs/store/Upload.spec.js
For manual testing, create an input element on UploadView component and manually upload files from computer and upload it.

## Screenshots (if appropriate):
id201
id202

## Types of changes

- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/store/modules/upload/index.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    // pre-set upload status
+    context.commit(SET_IS_UPLOADING, true)
+    context.commit(SET_BYTES_UPLOADED, 0)
+    context.commit(SET_BYTES_REMAINING, file.size)
+    // start upload
+    var upload = storageDB.child('files/' + file.name).put(file)
+    upload.on(Firebase.storage.TaskEvent.STATE_CHANGED, function (snapshot) {
+      // update upload status
+      context.commit(SET_BYTES_UPLOADED, snapshot.bytesTransferred)
+      context.commit(SET_BYTES_REMAINING, snapshot.totalBytes - snapshot.bytesTransferred)
```
Comment (peggyopal) [inline src/store/modules/upload/index.js]: Lines 53-55 are indented too much

File: test/unit/specs/UploadView.spec.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  it('should render correct contents', () => {
+    const Constructor = Vue.extend(Upload)
+    const vm = new Constructor().$mount()
+    const input = vm.$el.getElementsByTagName('input')[0]
+    expect(input.type)
+      .to.equal('file')
+    expect(input.accept)
+      .to.equal('image/*')
+  })
+})
```
Comment (peggyopal) [inline test/unit/specs/UploadView.spec.js]: These changes shouldn't be in this PR

File: test/unit/specs/store/Upload.spec.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+const testAction = (action, args, state, expectedMutations, done) => {
+  let count = 0
+
+  // mock commit
+  const commit = (type, payload) => {
+    const mutation = expectedMutations[count]
+
+    try {
+      expect(mutation.type).to.equal(type)
```
Comment (peggyopal) [inline test/unit/specs/store/Upload.spec.js]: what exactly is this doing?

File: test/unit/specs/store/Upload.spec.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      { type: 'SET_BYTES_REMAINING', payload: 100 },
+      { type: 'SET_BYTES_UPLOADED', payload: 100 },
+      { type: 'SET_BYTES_REMAINING', payload: 0 },
+      { type: 'SET_IS_UPLOADING', payload: false },
+      { type: 'SET_BYTES_UPLOADED', payload: 0 },
+      { type: 'SET_BYTES_REMAINING', payload: 0 },
+      { type: 'SET_CANCEL_UPLOAD', payload: false }
+    ], done)
+  })
+})
```
Comment (peggyopal) [inline test/unit/specs/store/Upload.spec.js]: Do you have any tests for upload failures?

File: src/components/UploadView.vue [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+      this.files = tmpArray.filter(file => file.type.indexOf('image/') !== -1)
+      this.$store.dispatch('uploadSingleFile', this.files[0])
+    }
+  }
+}
+</script>
+
+<style scoped>
+
+</style>
```
[the remark this answers is not in the document]
Comment (MoranLi) [reply src/components/UploadView.vue]: this vue is for testing functionality of upload action, but I can remove It

File: test/unit/specs/store/Upload.spec.js (hunk 3 above)
Comment (MoranLi) [reply test/unit/specs/store/Upload.spec.js]: the function will go through each mutation call by action and compare with the exceptedMutations

File: src/components/UploadView.vue (hunk 5 above)
[the remark this answers is not in the document]
Comment (peggyopal) [reply src/components/UploadView.vue]: I may have gotten my files wrong then... If you have tests for front-end work then they shouldn't be in here

File: test/unit/specs/store/Upload.spec.js (hunk 4 above)
Comment (MoranLi) [reply test/unit/specs/store/Upload.spec.js]: only way cause a failure right not is shut down wifi

File: test/unit/specs/UploadView.spec.js (hunk 2 above)
Comment (MoranLi) [reply test/unit/specs/UploadView.spec.js]: this is the testing file for the ui, but I can remove it

File: src/store/modules/upload/index.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+// getters for this module's state
+const getters = {
+  getIsUploading: state => state.isUploading,
+  getBytesUploaded: state => state.bytesUploaded,
+  getBytesRemaining: state => state.bytesRemaining,
+  getCancelUpload: state => state.cancelUpload
+}
+
+// mutations of this module, mutation must be sync and atomic
```
[the remark this answers is not in the document]
Comment (MoranLi) [reply src/store/modules/upload/index.js]: sync and atomic

File: test/unit/specs/store/Upload.spec.js (hunk 3 above)
Comment (peggyopal) [reply test/unit/specs/store/Upload.spec.js]: Sweet! Thanks :)

File: test/unit/specs/store/Upload.spec.js (hunk 4 above)
Comment (peggyopal) [reply test/unit/specs/store/Upload.spec.js]: Awesome! Thanks!!

File: src/store/modules/upload/index.js (hunk 1 above)
Comment (MoranLi) [reply src/store/modules/upload/index.js]: it`s a function which will call back by upload.on(), or if my format is wrong that callback function should not write in this way?

File: test/unit/specs/store/Upload.spec.js (hunk 4 above)
Comment (MoranLi) [reply test/unit/specs/store/Upload.spec.js]: the error handling is actually done by firebase itself, so sorry what I meant by in the last comment is that shut down wifi is the only way that I can fail the upload by my self...

File: src/store/modules/upload/index.js (hunk 1 above)
Comment (peggyopal) [reply src/store/modules/upload/index.js]: My bad I did not catch that. You're good!

File: src/components/UploadView.vue (hunk 5 above)
[the remark this answers is not in the document]
Comment (KingMackle) [reply src/components/UploadView.vue]: Yes Li, there is no problem having a dummy page to test your backend code locally, but on the main code stream we would like to have it clean of any prototyping stuff (ie. test page/component).

File: src/router/index.js [hunk 7]
```diff
@@ -61,6 +62,16 @@ const router = new Router({
...
+    },
+    {
+      path: '/upload',
+      name: 'Upload',
+      icon: 'event',
+      scrollToTop: true,
+      component: Upload,
+      meta: {
+        requiresAuth: true
+      }
```
[the remark this answers is not in the document]
Comment (KingMackle) [reply src/router/index.js]: Yes Peggy, you are correct. These changes shouldn't be part of this PR and shouldn't be pushed to the remote branches either.

File: src/router/index.js [hunk 8]
```diff
@@ -8,6 +8,7 @@ const Slides = () => import(/* webpackChunkName: "Slides.vue" */'@/components/Sl
 const Designer = () => import(/* webpackChunkName: "Designer.vue" */'@/components/Designer')
 const Calendar = () => import(/* webpackChunkName: "Calendar.vue" */'@/components/Calendar')
 /* const Login = () => import(/* webpackChunkName: "login.vue" \*\/ '@/components/Login') */
+const Upload = () => import(/* webpackChunkName: "Upload.vue" */'@/components/UploadView')
```
Comment (KingMackle) [inline src/router/index.js]: This should also be removed.

File: src/store/index.js [hunk 9]
```diff
@@ -15,6 +16,7 @@ export const store = new Vuex.Store({
   modules: {
     root,
     user,
+    Upload,
```
Comment (KingMackle) [inline src/store/index.js]: @MoranLi Can you please add this Upload module, into the hot-reloading section on line 27, 32, and 38 of this file. You can see how the slides module is imported on line 27, 32, and 38 of this file.

Comment (MoranLi) [reply src/store/index.js]: really sorry forget add that, should be fine now, the sample upload page should be removed now

**Pull request merged**